drivers/sensor/mhz19b/mhz19b.c

```diff
@@ -0,0 +1,351 @@
+	[MHZ19B_CMD_IDX_GET_CO2] = {
+		MHZ19B_HEADER, MHZ19B_RESERVED, MHZ19B_CMD_GET_CO2, MHZ19B_NULL_COUNT(5), 0x79
+	},
+	[MHZ19B_CMD_IDX_GET_RANGE] = {
+		MHZ19B_HEADER, MHZ19B_RESERVED, MHZ19B_CMD_GET_RANGE, MHZ19B_NULL_COUNT(5), 0x64
+	},
+	[MHZ19B_CMD_IDX_GET_ABC] = {
+		MHZ19B_HEADER, MHZ19B_RESERVED, MHZ19B_CMD_GET_ABC, MHZ19B_NULL_COUNT(5), 0x82
+	},
+	[MHZ19B_CMD_IDX_SET_ABC_ON] = {
+		MHZ19B_HEADER, MHZ19B_RESERVED, MHZ19B_CMD_SET_ABC, MHZ19B_ABC_ON,
+		MHZ19B_NULL_COUNT(4), 0xE6
+	},
+	[MHZ19B_CMD_IDX_SET_ABC_OFF] = {
+		MHZ19B_HEADER, MHZ19B_RESERVED, MHZ19B_CMD_SET_ABC, MHZ19B_ABC_OFF,
+		MHZ19B_NULL_COUNT(4), 0x86
+	},
+	[MHZ19B_CMD_IDX_SET_RANGE_2000] = {
+		MHZ19B_HEADER, MHZ19B_RESERVED, MHZ19B_CMD_SET_RANGE, MHZ19B_NULL_COUNT(3),
+		MHZ19B_RANGE_2000, 0x8F
+	},
+	[MHZ19B_CMD_IDX_SET_RANGE_5000] = {
+		MHZ19B_HEADER, MHZ19B_RESERVED, MHZ19B_CMD_SET_RANGE, MHZ19B_NULL_COUNT(3),
+		MHZ19B_RANGE_5000, 0xCB
+	},
+	[MHZ19B_CMD_IDX_SET_RANGE_10000] = {
+		MHZ19B_HEADER, MHZ19B_RESERVED, MHZ19B_CMD_SET_RANGE, MHZ19B_NULL_COUNT(3),
+		MHZ19B_RANGE_10000, 0x2F
+	},
+};
+
+static void mhz19b_uart_flush(const struct device *uart_dev)
+{
+	uint8_t c;
+
+	while (uart_fifo_read(uart_dev, &c, 1) > 0) {
+		continue;
+	}
+}
+
+static uint8_t mhz19b_checksum(const uint8_t *data)
+{
+	uint8_t cs = 0;
+
+	for (uint8_t i = 1; i < MHZ19B_BUF_LEN - 1; i++) {
+		cs += data[i];
+	}
+
+	return 0xff - cs + 1;
+}
+
+static int mhz19b_send_cmd(const struct device *dev, const enum mhz19b_cmd_idx cmd_idx,
+			   const bool has_rsp)
+{
+	struct mhz19b_data *data = dev->data;
+	const struct mhz19b_cfg *cfg = dev->config;
+	int ret;
+
+	/* Make sure last command has been transferred */
+	ret = k_sem_take(&data->tx_sem, MHZ19B_WAIT);
+	if (ret) {
+		return ret;
+	}
+
+	data->cmd_idx = cmd_idx;
+	data->has_rsp = has_rsp;
+	k_sem_reset(&data->rx_sem);
+
+	uart_irq_tx_enable(cfg->uart_dev);
+
+	if (has_rsp) {
+		uart_irq_rx_enable(cfg->uart_dev);
+		ret = k_sem_take(&data->rx_sem, MHZ19B_WAIT);
+	}
+
+	return ret;
+}
+
+static inline int mhz19b_send_config(const struct device *dev, const enum mhz19b_cmd_idx cmd_idx)
+{
+	struct mhz19b_data *data = dev->data;
+	int ret;
+
+	ret = mhz19b_send_cmd(dev, cmd_idx, true);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (data->rd_data[MHZ19B_RX_CMD_IDX] != mhz19b_cmds[data->cmd_idx][MHZ19B_TX_CMD_IDX]) {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static inline int mhz19b_poll_data(const struct device *dev, const enum mhz19b_cmd_idx cmd_idx)
+{
+	struct mhz19b_data *data = dev->data;
+	uint8_t checksum;
+	int ret;
+
+	ret = mhz19b_send_cmd(dev, cmd_idx, true);
+	if (ret < 0) {
+		return ret;
+	}
+
+	checksum = mhz19b_checksum(data->rd_data);
+	if (checksum != data->rd_data[MHZ19B_CHECKSUM_IDX]) {
+		LOG_DBG("Checksum mismatch: 0x%x != 0x%x", checksum,
+			data->rd_data[MHZ19B_CHECKSUM_IDX]);
+		data->data_valid = false;
+		return -EBADMSG;
+	}
+
+	data->data_valid = true;
```

Comment:
data_valid can now be deleted (not used)

Reply:
Thanks